### ISSUE TITLE
[2.0] Keep-alive causes error in transitioned router-view 

### ISSUE THREAD

**Issue by fnlctrl** [vuejs/vue-router#615]
### Vue.js & vue-router.js version

2.0.0-rc.2, 2.0.0-rc.3
### Steps to reproduce

Click on any of the router-links
### What is Expected?

Without keep-alive everything works fine (current router-view fades, then next router-view appears)
https://jsfiddle.net/fnlCtrl/jLgy5ma4/
### What is actually happening?

With keep-alive:
https://jsfiddle.net/fnlCtrl/9nnxaea5/3/
Bug in which the update is applied to both enter and leave element, their contents become the same (current router-view becomes next router-view, then it fades, then next router-view appears)

**Comment by yyx990803** [comment]
This is not really an error per-se, because with keep-alive, the faded out instance is, well, kept alive... so it would still react to data changes (in this case, the route param) while it's being faded out.

**Comment by fnlctrl** [comment]
Well... so the only workaround would be somehow freezing the leaving component. I'll try to figure that out today...
Meanwhile, I guess this will be a major problem for people using `transition` + `keep-alive` when both of the entering and leaving components use the same data source. So.. I guess it should be addressed in the docs at least.

**Comment by fnlctrl** [comment]
I've encountered another problem using keep-alive with router-view. (It's not directly about this issue, but it's somehow related.)
Let's say we have a route map:

```
| - /foo [Foo] (kept-alive)
  | - [Bar] (inside Foo's template, which adds a scroll listener on window)
```

When Foo deactivates, it's `deactivated` hook gets called, we can use it to notify Bar that its parent is deactivated, and tell Bar to remove the `scroll` listener it added on window.

But notifying `parent deactivated` to children can be cumbersome when we have nested components

```
| - /foo [Foo] (kept-alive)
  | [Baz] (inside Foo's template)
    | [Bar] (inside Baz's template)
```

When Foo deactivates, it'll have to notify Baz, and then Baz will have to notify Bar to remove the listeners.

My current work around to this problem is to add a `routeActive` property to Foo, and then add `routeActive` computed property (return this.$parent.routeActive) on Baz and Bar. Though it has one limitation: `routeActive` as a computed property cannot be watched (because it relies on this.$parent). 
So I'll have to add a `if (!this.routeActive) return` check on event listener and watchers (for example, on `$route`) of Baz and Bar. 

**To summarize, the problem is actually about notifying a kept-alive component's children when their parent gets activated/deactivated.**

Currently I came up with two possible solutions:
1. An option to `freeze` the kept-alive component and its children when it deactivates (which solves the original issue about transition with keep-alive, though I guess it'll be very hard to implement),
   but we still have to manually tell children components to remove listeners added to window.
2. Add a `parentDeactivated` hook option for child components. Though there may be performance issues...

@yyx990803 What's your suggestion about this?

**Comment by LinusBorg** [comment]
@fnlctrl Is this still an issue?

**Comment by fnlctrl** [comment]
Yeah, but I've already worked around this. Seems no one else is having this
issue so I guess it's safe to close it for now..
